### PR TITLE
Implement Process.Modules on OS X

### LINKED ISSUE #1727: Investigate implementing Process.Modules on Unix
On Unix, we currently return an empty array from Process.Modules, and we throw a PlatformNotSupportedException from Process.MainModule.  Is there something better we can reasonably do on Linux and OSX?  If so, we should do it.

### PR DESCRIPTION
**Pull request by stephentoub**
Today Process.Modules just returns an empty collection on OS X, whereas on Linux we use data from procfs to populate it.  This commit adds similar support on OS X by popen'ing vmmap.

cc: @Priya91, @pallavit, @sokket 
Fixes #1727

### CODE REVIEW

File: src/System.Diagnostics.Process/src/System/Diagnostics/ProcessManager.OSX.cs [hunk 1]
```diff
@@ -103,8 +105,76 @@ private static ProcessInfo CreateProcessInfo(int pid)
...
+                            continue;
+                        }
+
+                        // Find the first space-slash.  If it exists, that's the start of the path.
+                        int filePathPos = line.IndexOf(PathStart, addressEndPos);
+                        if (filePathPos < 0)
+                        {
+                            continue;
+                        }
+                        string filePath = line.Substring(filePathPos).Trim();
```
Comment (justinvp) [inline src/System.Diagnostics.Process/src/System/Diagnostics/ProcessManager.OSX.cs]: `SubstringTrim`?

Comment (stephentoub) [inline src/System.Diagnostics.Process/src/System/Diagnostics/ProcessManager.OSX.cs]: Done.

File: src/System.Diagnostics.Process/src/System/Diagnostics/ProcessManager.OSX.cs [hunk 2]
```diff
@@ -103,8 +105,76 @@ private static ProcessInfo CreateProcessInfo(int pid)
...
+
+            // On Linux we can read from procfs.  On OS X we can get similar data by getting the output from vmmap.
+            // This is the same basic approach taken by the PAL in libcoreclr.
+            IntPtr popenStdout = Interop.Sys.POpen($"/usr/bin/vmmap -interleaved {processId} -wide", "r");
+            Debug.Assert(popenStdout != IntPtr.Zero, $"popen failed: {Interop.Sys.GetLastErrorInfo()}");
+            if (popenStdout != IntPtr.Zero) // if this failed, we'll just return an empty set of modules
+            {
+                try
+                {
+                    string line;
```
Comment (sokket) [inline src/System.Diagnostics.Process/src/System/Diagnostics/ProcessManager.OSX.cs]: Your comment specifies that line should be free'd by the caller but I don't see any Free calls...does line need to be free'd?

**Comment by sokket** [comment]
One question but otherwise LGTM

File: src/System.Diagnostics.Process/src/System/Diagnostics/ProcessManager.OSX.cs (hunk 2 above)
Comment (stephentoub) [inline src/System.Diagnostics.Process/src/System/Diagnostics/ProcessManager.OSX.cs]: It's done automatically by the marshaler when going from the returned char* to a string.

Comment (sokket) [inline src/System.Diagnostics.Process/src/System/Diagnostics/ProcessManager.OSX.cs]: That's fancy...didn't know that. Thanks!

**Comment by Priya91** [comment]
LGTM as well !

**Comment by Priya91** [comment]
@dotnet-bot test Innerloop OSX Debug Build and Test

**Comment by stephentoub** [comment]
Thanks for the review!

**Comment by stephentoub** [comment]
Test Innerloop OSX Debug Build and Test please

(The second run passed, but I want to run it again one more time, as it looks like the first run hung for some reason.)

**Pull request merged**